### PR TITLE
Goa 2943 distinct facet count in statistics

### PR DESCRIPTION
**Pull request by TonyWardell**
Achieve this by adding numBuckets:true to get agg_facet request

Changes response from solr to 
   "unique_geneProductId": 41458221,
    "agg_evidenceCode": {
      "numBuckets": 35,
      "buckets": [
        {
          "val": "ECO:0000256",
          "count": 88001168,
          "unique_geneProductId": 38321802,
          "count_id": 88001168
        },
        {
          "val": "ECO:0000323",
          "count": 50206450,
          "unique_geneProductId": 25361638,
          "count_id": 50206450
        },
 ...

And the resulting json from the rest service
  "groupName": "annotation",
      "totalHits": 27,
      "types": [
        {
          "type": "reference",
          "distinctValueCount": 9,
          "values": [
            {
              "key": "PMID:25911094",
              "percentage": 0.4074074074074074,
              "hits": 11,
              "name": null
            },
            {
              "key": "GO_REF:0000037",
              "percentage": 0.14814814814814814,
              "hits": 4,
              "name": null
            },
            {
              "key": "GO_REF:0000107",
              "percentage": 0.14814814814814814,
              "hits": 4,
              "name": null
            },
            {
              "key": "GO_REF:0000039",
              "percentage": 0.1111111111111111,
              "hits": 3,
              "name": null
            },
            {
              "key": "PMID:14647039",
              "percentage": 0.037037037037037035,
              "hits": 1,
              "name": null
            },
            {
              "key": "PMID:16674676",
              "percentage": 0.037037037037037035,
              "hits": 1,
              "name": null
            },
            {
              "key": "Reactome:R-HSA-392499",
              "percentage": 0.037037037037037035,
              "hits": 1,
              "name": null
            },
            {
              "key": "Reactome:R-HSA-976734",
              "percentage": 0.037037037037037035,
              "hits": 1,
              "name": null
            },
            {
              "key": "Reactome:R-HSA-977136",
              "percentage": 0.037037037037037035,
              "hits": 1,
              "name": null
            }
          ]

### CODE REVIEW

File: rest-common/src/test/java/uk/ac/ebi/quickgo/rest/search/solr/SolrResponseAggregationConverterTest.java [hunk 1]
```diff
@@ -118,6 +119,28 @@ public void solrResponseWithNonAggregatedBucketInGlobalAggregateConvertsToAEmpty
         assertThat(agg.getNestedAggregations(), hasSize(0));
     }
 
+    @Test
+    public void solrResponseWithTotalDistinctCountForFacetIsAddedToNestedAggregation() throws Exception {
```
Comment (eddturner) [inline rest-common/src/test/java/uk/ac/ebi/quickgo/rest/search/solr/SolrResponseAggregationConverterTest.java]: I was looking for tests in AnnotationControllerStatisticsIT to easily see the effects of this. It would be good to add some there to show it's effects.

Comment (TonyWardell) [reply rest-common/src/test/java/uk/ac/ebi/quickgo/rest/search/solr/SolrResponseAggregationConverterTest.java]: Done

File: rest-common/src/test/java/uk/ac/ebi/quickgo/rest/search/solr/SolrAggregationHelperTest.java [hunk 2]
```diff
@@ -78,47 +78,56 @@ public void populatedTypeCreatesAggregateTypeTitle() throws Exception {
...
-    public void fieldWithNoPrefixReturnsAnEmptyPrefixWhenExtratingPrefix() throws Exception {
+    public void fieldWithNoPrefixReturnsAnEmptyPrefixWhenExtractingPrefix() throws Exception {
         String prefix = fieldPrefixExtractor(GP_ID_FIELD);
 
         assertThat(prefix, isEmptyString());
     }
 
     @Test
-    public void fieldWithCountPrefixReturnsAnCountPrefixWhenExtratingPrefix() throws Exception {
+    public void fieldWithCountPrefixReturnsAnCountPrefixWhenExtractingPrefix() throws Exception {
```
[the remark this answers is not in the document]
Comment (TonyWardell) [reply rest-common/src/test/java/uk/ac/ebi/quickgo/rest/search/solr/SolrAggregationHelperTest.java]: Done

**Comment by TonyWardell** [comment]
Now with add rounding of the distinct value count for values greater than 10,000, to the nearest 10,000.

File: annotation-rest/src/main/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByType.java [hunk 3]
```diff
@@ -12,18 +14,24 @@
  */
 public class StatisticsByType {
     private final String type;
+    private final int distinctValueCount;
     private final List<StatisticsValue> values;
 
-    public StatisticsByType(String type) {
+    public StatisticsByType(String type, int distinctValueCount) {
         Preconditions.checkArgument(type != null && !type.isEmpty(), "Statistics type cannot be null or empty");
```
Comment (eddturner) [inline annotation-rest/src/main/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByType.java]: sanity precondition check + test, e.g., must be >= 0

File: annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByTypeTest.java [hunk 4]
```diff
@@ -42,18 +42,55 @@ public void nullStatisticsValueThrowsException() {
...
+    @Test
+    public void addedStatisticsValueRoundedIfValueGreaterThan10K() {
+        String type = "type";
+        StatisticsByType statsType = new StatisticsByType(type, 10001);
+
+        assertThat(statsType.getDistinctValueCount(), is(10000));
+    }
+
+    @Test
+    public void addedStatisticsValueRoundedIfValueGreaterThan10KTest2() {
```
Comment (eddturner) [inline annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByTypeTest.java]: either: addedStatisticsValueRoundedIfValueGreaterThan10KTest2 -> distinctValueCountRoundedUpWhenGreaterThan10K

or the test can be removed, since I think is made redundant by a test lower down

File: annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByTypeTest.java [hunk 5]
```diff
@@ -42,18 +42,55 @@ public void nullStatisticsValueThrowsException() {
...
-        StatisticsValue value = new StatisticsValue("key", 0, 0);
+        StatisticsByType statsType = new StatisticsByType(type, 12);
 
-        StatisticsByType statsType = new StatisticsByType(type);
-        statsType.addValue(value);
+        assertThat(statsType.getDistinctValueCount(), is(12));
+    }
+
+    @Test
+    public void addedStatisticsValueRoundedIfValueGreaterThan10K() {
```
Comment (eddturner) [inline annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByTypeTest.java]: addedStatisticsValueRoundedIfValueGreaterThan10K -> distinctValueCountRoundedDownWhenGreaterThan10K

File: annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByTypeTest.java [hunk 6]
```diff
@@ -42,18 +42,55 @@ public void nullStatisticsValueThrowsException() {
         thrown.expect(IllegalArgumentException.class);
         thrown.expectMessage("Stats value cannot be null");
 
-        StatisticsByType statsType = new StatisticsByType(type);
+        StatisticsByType statsType = new StatisticsByType(type, 0);
         statsType.addValue(value);
     }
```
Comment (eddturner) [inline annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByTypeTest.java]: need a test case for the boundary condition of when the distinct count == 0 (even though it's a trivial answer)

File: annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByTypeTest.java [hunk 7]
```diff
@@ -42,18 +42,55 @@ public void nullStatisticsValueThrowsException() {
...
+    @Test
+    public void roundUpDistinctCount() {
+        String type = "type";
+        StatisticsByType statsType = new StatisticsByType(type, 487934);
+
+        assertThat(statsType.getDistinctValueCount(), is(488000));
+    }
+
+    @Test
+    public void noNeedToRoundUpDistinctCount() {
```
Comment (eddturner) [inline annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByTypeTest.java]: noNeedToRoundUpDistinctCount -> roundDownDistinctCount, just to make the test clear rounding is still expected

File: annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/service/converter/WorkbookFromStatisticsImplTest.java [hunk 8]
```diff
@@ -121,25 +121,25 @@ private void buildStatisticGroups() {
...
         //GoId
-        StatisticsByType statisticsByAnnotationTypeGoId = new StatisticsByType("TEST_GO_ID");
+        StatisticsByType statisticsByAnnotationTypeGoId = new StatisticsByType("TEST_GO_ID",0);
         statisticsByAnnotationTypeGoId.addValue(new StatisticsValue("GO:0003824", 1, 5L));
         statisticsByAnnotationTypeGoId.addValue(new StatisticsValue("GO:0003870",1, 5L));
         statisticsByAnnotationTypeGoId.addValue(new StatisticsValue("GO:0009058",1, 5L));
         annotationStatisticsGroup.addStatsType(statisticsByAnnotationTypeGoId);
 
-        StatisticsByType statisticsByGeneProductTypeGoId = new StatisticsByType("TEST_GO_ID");
+        StatisticsByType statisticsByGeneProductTypeGoId = new StatisticsByType("TEST_GO_ID",0);
```
Comment (eddturner) [inline annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/service/converter/WorkbookFromStatisticsImplTest.java]: reformatting required on this class (i just see it now, sorry!)

File: annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByTypeTest.java [hunk 9]
```diff
@@ -42,18 +42,55 @@ public void nullStatisticsValueThrowsException() {
...
+        String type = "type";
+        StatisticsByType statsType = new StatisticsByType(type, 333434);
+
+        assertThat(statsType.getDistinctValueCount(), is(333000));
+    }
+
+    @Test
+    public void distinctCountGreaterThan1m() {
```
Comment (eddturner) [inline annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByTypeTest.java]: test not necessary (covered by the other rounding up/down ones), but no prob to leave in -- however, to describe intention of the test, think it should include the up/down part of rounding.

e.g., roundDownDistinctCountGreaterThan1m

File: annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/controller/AnnotationControllerStatisticsIT.java [hunk 10]
```diff
@@ -104,7 +106,7 @@ public void queryWithNoHitsProducesEmptyStats() throws Exception {
     //----------- Ontology ID -----------//
     @Test
     public void statsForAllDocsContaining1OntologyIdReturns1OntologyIdStat() throws Exception {
-        executesAndAssertsCalculatedStatsForAttribute(GO_ID_STATS_FIELD, savedDocs, doc -> doc.goId);
+        executesAndAssertsCalculatedStatsForAttribute(GO_ID_STATS_FIELD, savedDocs, doc -> doc.goId, 1);
```
Comment (eddturner) [inline annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/controller/AnnotationControllerStatisticsIT.java]: cool

File: annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByTypeTest.java [hunk 11]
```diff
@@ -42,18 +42,55 @@ public void nullStatisticsValueThrowsException() {
...
+    @Test
+    public void addedStatisticsValueRoundedIfValueGreaterThan10KTest2() {
+        String type = "type";
+        StatisticsByType statsType = new StatisticsByType(type, 11999);
+
+        assertThat(statsType.getDistinctValueCount(), is(12000));
+    }
+
+    @Test
+    public void roundUpDistinctCount() {
```
Comment (eddturner) [inline annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByTypeTest.java]: I think the tested value should be a boundary case, e.g., 487500, just for peace of mind (since this is when rounding up should start from from what I understand)

File: annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByTypeTest.java (hunk 7 above)
Comment (eddturner) [reply annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByTypeTest.java]: I think the tested value should be a boundary case, e.g., 333499, (so we don't need to add other explicit boundary test cases)

File: annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByTypeTest.java [hunk 12]
```diff
@@ -42,18 +42,55 @@ public void nullStatisticsValueThrowsException() {
         thrown.expect(IllegalArgumentException.class);
         thrown.expectMessage("Stats value cannot be null");
 
-        StatisticsByType statsType = new StatisticsByType(type);
+        StatisticsByType statsType = new StatisticsByType(type, 0);
         statsType.addValue(value);
     }
 
     @Test
     public void addedStatisticsValueIsRetrievedCorrectly() {
```
Comment (eddturner) [inline annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByTypeTest.java]: addedStatisticsValueIsRetrievedCorrectly -> distinctValueCountRetrievedExactlyWhenUnder10000

... and to test the boundary, if this value was changed from 12 to 9999, we'd wouldn't need another explicit test testing this boundary

File: annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByTypeTest.java [hunk 13]
```diff
@@ -42,18 +42,55 @@ public void nullStatisticsValueThrowsException() {
...
         String type = "type";
-        StatisticsValue value = new StatisticsValue("key", 0, 0);
+        StatisticsByType statsType = new StatisticsByType(type, 12);
 
-        StatisticsByType statsType = new StatisticsByType(type);
-        statsType.addValue(value);
+        assertThat(statsType.getDistinctValueCount(), is(12));
+    }
+
```
Comment (eddturner) [inline annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByTypeTest.java]: one more test to test what happens if the distinct value is 10000 exactly to test that boundary

File: annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsGroupTest.java [hunk 14]
```diff
@@ -66,7 +66,7 @@ public void addedStatisticsTypeIsRetrievedCorrectly() {
         String groupName = "group";
         int totalHits = 0;
 
-        StatisticsByType statsType = new StatisticsByType("type");
+        StatisticsByType statsType = new StatisticsByType("type",0);
```
Comment (eddturner) [inline annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsGroupTest.java]: formatting

File: annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByTypeTest.java (hunk 6 above)
Comment (TonyWardell) [reply annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByTypeTest.java]: Done

File: annotation-rest/src/main/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByType.java (hunk 3 above)
Comment (TonyWardell) [reply annotation-rest/src/main/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByType.java]: Done

File: annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByTypeTest.java (hunk 12 above)
Comment (TonyWardell) [reply annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByTypeTest.java]: the boundry is really between 10000 & 10001 (at least in terms of discernible difference of output) so I've change the class and test to match.

File: annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByTypeTest.java (hunk 13 above)
Comment (TonyWardell) [reply annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByTypeTest.java]: Changing value tested in StatisticsByCount changes the boundry, so now either side of the boundary is tested.

File: annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByTypeTest.java (hunk 5 above)
Comment (TonyWardell) [reply annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByTypeTest.java]: Names changed.

File: annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByTypeTest.java (hunk 11 above)
Comment (TonyWardell) [reply annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByTypeTest.java]: Agreed, (in fact I preempted this comment before I saw it) and hopefully its clearer now.

File: annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByTypeTest.java (hunk 9 above)
Comment (TonyWardell) [reply annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByTypeTest.java]: I will leave these tests in as it makes it clear that rounding up and down occurs beyond 10500

File: annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByTypeTest.java (hunk 7 above)
Comment (TonyWardell) [reply annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByTypeTest.java]: Tests changed

File: annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByTypeTest.java (hunk 9 above)
Comment (TonyWardell) [reply annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsByTypeTest.java]: Removed.

File: annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsGroupTest.java (hunk 14 above)
Comment (TonyWardell) [reply annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/model/StatisticsGroupTest.java]: Done

File: annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/service/converter/WorkbookFromStatisticsImplTest.java (hunk 8 above)
Comment (TonyWardell) [reply annotation-rest/src/test/java/uk/ac/ebi/quickgo/annotation/service/converter/WorkbookFromStatisticsImplTest.java]: Done

File: rest-common/src/main/java/uk/ac/ebi/quickgo/rest/search/solr/SolrResponseAggregationConverter.java [hunk 15]
```diff
@@ -121,6 +122,9 @@ private void convertAggregateValue(String field, Object value, AggregateResponse
             } else {
                 logger.debug("Unable to process field:{}, with prefix:{}", field, fieldPrefix);
             }
+        } else if (SolrAggregationHelper.distinctValueCountTester(field)){
+            aggregation.setDistinctValuesCount(((Number)value).intValue());
```
[the remark this answers is not in the document]
Comment (TonyWardell) [reply rest-common/src/main/java/uk/ac/ebi/quickgo/rest/search/solr/SolrResponseAggregationConverter.java]: Builder built.

File: rest-common/src/main/java/uk/ac/ebi/quickgo/rest/search/solr/SolrAggregationHelper.java [hunk 16]
```diff
@@ -70,4 +72,13 @@ static String fieldNameExtractor(String field) {
...
     }
-}
\ No newline at end of file
+
+    /**
+     * Test the field to see if holds the result of the distinct value count.
+     * @param field element of the aggregation result
+     * @return true if field holds distinct value count field definition.
+     */
+    static boolean distinctValueCountTester(String field) {
```
[the remark this answers is not in the document]
Comment (TonyWardell) [reply rest-common/src/main/java/uk/ac/ebi/quickgo/rest/search/solr/SolrAggregationHelper.java]: Done
